lib/bundling.js

```diff
@@ -0,0 +1,256 @@
+        };
+      }
+    })
+    .filter(match => match)
+    .map(style => {
+      const name = style.name.trim();
+      if (name.startsWith('@') && name.includes(' ')) {
+        const mediaQueryStyles = replaceSelectors(style.content, callback);
+        return `${style.name} {${mediaQueryStyles}}`;
+      } else {
+        return `${callback(style.selectors)} {${style.content}}`;
+      }
+    }).join('\n');
+}
+
+
+function scopeCSS(path, blockContent, $) {
+  const styles = replaceSelectors(blockContent, (selectors) => {
+    const scopedClass = `.b-${hash(removeExtension(path))}`;
+
+    return selectors.map(origSelector => {
+      let hasDeepCombinator = false;
+
+      return origSelector.reduce((selector, part) => {
+        if (!hasDeepCombinator && part === deepCombinator) {
+          hasDeepCombinator = true;
+          return selector;
+        }
+
+        if (hasDeepCombinator || combinators.includes(part) || part.startsWith(':')) {
+          // this part is not a queryable element, so it doesn't need a scoped css class
+          return `${selector} ${part}`;
+        }
+
+        const el = `${selector} ${part.replace(psuedoElements, '')}`;
+
+        if ($(el)) {
+          $(el).addClass(scopedClass.substring(1));
+        }
+
+        return `${selector} ${part.replace(/([^:]+)(:.+)?/, `$1${scopedClass}$2`)}`;
+      }, '').trim();
+    }).join(',\n');
+  });
+
+  return styles;
+}
+
+
+function writeEntryFile(type, blocksDir) {
+  const { bundles, ext } = blockTypes[type];
+  Object.keys(bundles).forEach((bundle) => {
+    fs.writeFileSync(`${blocksDir}/${bundle}.${ext}`, bundles[bundle].join('\n'));
+  });
+}
+
+
+function fixPaths(path, block, pathsRegex, root, blocksDir) {
+  return block.replace(pathsRegex, (match, _, importPath) => {
+    const abImportPath = resolve(root, dirname(path), importPath);
+    const newImportPath = relative(blocksDir, abImportPath);
+    return match.replace(importPath, newImportPath);
+  });
+}
+
+
+function extractBlocks(path, root, blocksDir) {
+  const blocks = {};
+  const contents = fs.readFileSync(path, 'utf8');
+  const template = /<template>[\s\S]*?<\/template>/gm.test(contents)
+    ? contents
+    : `<template>${contents}</template>`;
+  const original$ = vdom(template);
+
+  Object.entries(blockTypes).forEach(([ type, blockType ]) => {
+    const { tag, ext, pathsRegex } = blockType;
+
+    original$(tag).each((i, el) => {
+      const block = {
+        ext,
+        type,
+        file: getHashedPath(path, ext),
+        content: type === 'template'
+          ? minimizeWhitespace(original$)
+          : original$(el).get()[0].children[0].data,
+        ...el.attribs
+      };
+
+      if (type === 'js') {
+        addBundleImports(block, blockType);
+      }
+
+      if (type !== 'template') {
+        original$(el).remove();
+      }
+
+      block.content = fixPaths(path, block.content, pathsRegex, root, blocksDir);
```

Comment:
This is just an idea, but what about making `fixPaths` operate a little differently. This would allow you to pass in less argument variables. `fixPaths` could look like this:

```
function fixPaths(path, root, blocksDir)
  return (match, _, importPath) => {
    const abImportPath = resolve(root, dirname(path), importPath);
    const newImportPath = relative(blocksDir, abImportPath);
    return match.replace(importPath, newImportPath);
  };
}
```
And call it like this:
```
block.content = block.replace(pathsRegex, fixPaths(path, root, blocksDir));
```

Reply:
Interesting idea but this seems less clear and not worth it just to avoid one argument.

Reply:
It's two arguments, FYI, but 👍